Apps/TimelineApp/feature/user/src/main/java/uk/ac/cam/cares/jps/user/SensorSettingFragment.java

```diff
@@ -127,72 +138,88 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
         });
 
 
-        binding.topAppbar.setNavigationOnClickListener(view1 -> NavHostFragment.findNavController(this).navigateUp());
 
         binding.startRecordTv.setOnClickListener(this::onRecordButtonClicked);
+
+        binding.topAppbar.setNavigationOnClickListener(view1 -> NavHostFragment.findNavController(this).navigateUp());
+
         sensorViewModel.getIsRecording().observe(getViewLifecycleOwner(), isRecording -> {
             if (isRecording) {
-                binding.startRecordTv.setText(R.string.stop_recording);
-                adapter.setTogglesEnabled(false);
-                binding.toggleAllBtn.setEnabled(false);
+                stopRecording();
             } else {
-                binding.startRecordTv.setText(R.string.start_recording);
-                adapter.setTogglesEnabled(true);
-                updateStartRecordingButtonState();
-                resetToggleAllButton();
-                binding.toggleAllBtn.setEnabled(true);
+                startRecording();
             }
         });
     }
 
+
+
+    // Helper method to start recording the selected sensors
+    private void startRecording() {
+        binding.startRecordTv.setText(R.string.start_recording);
+        adapter.setTogglesEnabled(true);
+        resetToggleAllButton();
+        binding.toggleAllBtn.setEnabled(true);
+    }
+
+    // Helper method to stop recording and reset the sensors
+    private void stopRecording() {
+        binding.startRecordTv.setText(R.string.stop_recording);
+        adapter.setTogglesEnabled(false);
+        binding.toggleAllBtn.setEnabled(false);
+    }
+
     /**
      * Resets the state of the "Toggle All" button to its default state.
      */
     private void resetToggleAllButton() {
         binding.toggleAllBtn.setText(R.string.toggle_all);
     }
 
-    /**
-     * Updates the state of the "Start Recording" button based on the current sensor toggle states.
-     */
-    private void updateStartRecordingButtonState() {
-        // enable the start recording button only if at least one sensor is toggled on
-        boolean hasToggledOnSensor = false;
-        for (SensorItem item : adapter.getSensorItems()) {
-            if (item.isToggled()) {
-                hasToggledOnSensor = true;
-                break;
-            }
-        }
-        binding.startRecordTv.setEnabled(hasToggledOnSensor);
-    }
 
     /**
      * Handles the logic for starting or stopping sensor recording when the button is clicked.
      *
      * @param view The view that was clicked.
      */
     private void onRecordButtonClicked(View view) {
```

Comment:
The state management for this function is quite confusing. Please update this method
1. Probably OK: Gather all the selectedSensors from adapter, but will be more ideal to just use selectedSensors in ViewModel to make it the only source of true of state
    - selectedSensors is updated whenever a sensor is toggled
2. OK: Check whether location and audio have been toggled and request for the corresponding permission
3. Not Good: manually call startRecording(...). It should change `isRecording` state only and listen to this state to perform the corresponding UI updates and viewmodel function calls. And selectedSensors should not be passed into ViewModel, it should be set into ViewModel, and the ViewModel is the only source of true of state.
4. Not Good: Should not set listener and onClickeListener in this method unless they response differently every time the button is clicked. Instead they should be defined in onViewCreated